DefaultFileDownloader.executeViaProxy() not working with HTTP Proxy for HTTPS connections

I'm running a local CNTLM proxy to access the corporate ISA Proxy. So my Maven ``settings.xml`` holds the following proxy configuration:
```
	<proxies>
		<proxy>
			<id>cntlm-http</id>
			<protocol>http</protocol>
			<host>localhost</host>
			<port>3128</port>
		</proxy>
		<proxy>
			<id>cntlm-https</id>
			<protocol>https</protocol>
			<host>localhost</host>
			<port>3128</port>
		</proxy>
	</proxies>
```
Using this proxy configuration for installing node / npm from ``https://nodejs.org`` fails with the following error:
```
[INFO] --- frontend-maven-plugin:0.0.27:install-node-and-npm (install node and npm) @ webui ---
[INFO] Found proxies: [cntlm-http{protocol='http', host='localhost', port=3128}, cntlm-https{protocol='https', host='localhost', port=3128}]
[INFO] Installing node version v0.12.2
[INFO] Downloading via proxy cntlm-https{protocol='https', host='localhost', port=3128}
Jan 20, 2016 2:01:02 PM org.apache.http.impl.execchain.RetryExec execute
INFORMATION: I/O exception (java.net.SocketException) caught when processing request to {tls}->https://localhost:3128->https://nodejs.org:443: Connect
ion reset
Jan 20, 2016 2:01:02 PM org.apache.http.impl.execchain.RetryExec execute
INFORMATION: Retrying request to {tls}->https://localhost:3128->https://nodejs.org:443
Jan 20, 2016 2:01:02 PM org.apache.http.impl.execchain.RetryExec execute
INFORMATION: I/O exception (java.net.SocketException) caught when processing request to {tls}->https://localhost:3128->https://nodejs.org:443: Connect
ion reset
Jan 20, 2016 2:01:02 PM org.apache.http.impl.execchain.RetryExec execute
INFORMATION: Retrying request to {tls}->https://localhost:3128->https://nodejs.org:443
Jan 20, 2016 2:01:02 PM org.apache.http.impl.execchain.RetryExec execute
INFORMATION: I/O exception (java.net.SocketException) caught when processing request to {tls}->https://localhost:3128->https://nodejs.org:443: Connect
ion reset
Jan 20, 2016 2:01:02 PM org.apache.http.impl.execchain.RetryExec execute
INFORMATION: Retrying request to {tls}->https://localhost:3128->https://nodejs.org:443
```
The reason for this error is that the proxy drops the connection because it's accessed via HTTPS (``{tls}->https://localhost:3128->https://nodejs.org:443``):
```
Caused by: java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:209)
        at java.net.SocketInputStream.read(SocketInputStream.java:141)
        at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
        at sun.security.ssl.InputRecord.read(InputRecord.java:503)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:973)
        at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1375)
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1403)
        at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1387)
        at org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:394)
        at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:353)
        at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:134)
        at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
        at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:388)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
        at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
        at com.github.eirslett.maven.plugins.frontend.lib.DefaultFileDownloader.executeViaProxy(FileDownloader.java:111)
        at com.github.eirslett.maven.plugins.frontend.lib.DefaultFileDownloader.execute(FileDownloader.java:83)
        at com.github.eirslett.maven.plugins.frontend.lib.DefaultFileDownloader.download(FileDownloader.java:59)
```
The corresponding debug log from CNTLM is as following:
```
******* Round 1 C: 4 *******
Reading headers (4)...
HEAD: ♥♥headers_recv: Unknown header (♥♥).
headers_recv: fd 4 error -4

******* Round 1 C: 4 *******
Joining thread 537275336; rc: 0
Reading headers (4)...
HEAD: ♥♥headers_recv: Unknown header (♥♥).
headers_recv: fd 4 error -4

******* Round 1 C: 4 *******
Joining thread 537275480; rc: 0
Reading headers (4)...
HEAD: ♥♥headers_recv: Unknown header (♥♥).
headers_recv: fd 4 error -4

******* Round 1 C: 4 *******
Joining thread 537275336; rc: 0
Reading headers (4)...
HEAD: ♥♥headers_recv: Unknown header (♥♥).
headers_recv: fd 4 error -4
Joining thread 537275480; rc: 0
```
The culprit is the initialization of the proxy HttpHost with the proxy protocol in [``DefaultFileDownloader.executeViaProxy()``](https://github.com/eirslett/frontend-maven-plugin/blob/master/frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/FileDownloader.java#L104). This should be changed as follows:
```
final HttpHost proxyHttpHost = new HttpHost(proxy.host, proxy.port);
```

As a side note: [``ProxyConfig.getURI()``](https://github.com/eirslett/frontend-maven-plugin/blob/master/frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/ProxyConfig.java#L85) holds the following comment:
```
// Proxies should be schemed with http, even if the protocol is https
```

--- a/frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/FileDownloader.java
+++ b/frontend-plugin-core/src/main/java/com/github/eirslett/maven/plugins/frontend/lib/FileDownloader.java
@@ -101,7 +101,7 @@ final class DefaultFileDownloader implements FileDownloader {
             proxyClient = buildHttpClient(null);
         }
 
-        final HttpHost proxyHttpHost = new HttpHost(proxy.host, proxy.port, proxy.protocol);
+        final HttpHost proxyHttpHost = new HttpHost(proxy.host, proxy.port);
 
         final RequestConfig requestConfig = RequestConfig.custom().setProxy(proxyHttpHost).build();
 

